cmd/checkpoint/main.go

```diff
@@ -505,14 +545,85 @@ func (c *kubeletClient) localParentPods() map[string]*v1.Pod {
 	return podListToParentPods(podList)
 }
 
-// localRunningPods uses the /runningpods/ kubelet api to retrieve the local container runtime pod state
+// localRunningPods uses the CRI shim to retrieve the local container runtime pod state
 func (c *kubeletClient) localRunningPods() map[string]*v1.Pod {
-	podList := new(v1.PodList)
-	if err := c.secureClient.Get().AbsPath("/runningpods/").Do().Into(podList); err != nil {
-		// Assume there are no local running pods.
-		glog.Errorf("failed to list local running pods, assuming none are ready: %v", err)
+
+	pods := make(map[string]*v1.Pod)
+
+	sandboxes, err := c.getKubeletSandboxes(false)
+	if err != nil {
+		glog.Errorf("failed to list running sandboxes: %v", err)
+		return nil
+	}
+
+	// Add all pods that sandboxes are apart of
+	for _, s := range sandboxes {
+		if s.Metadata == nil {
+			glog.V(4).Infof("Sanbox does not have metadata: %+v", s)
+			continue
+		}
+
+		podName := s.Metadata.Namespace + "/" + s.Metadata.Name
+		if _, ok := pods[podName]; !ok {
+			p := &v1.Pod{}
+			p.UID = types.UID(s.Metadata.Uid)
+			p.Name = s.Metadata.Name
+			p.Namespace = s.Metadata.Namespace
+
+			pods[podName] = p
+		}
+	}
+
+	containers, err := c.getKubeletContainers(false)
```

Comment:
I like the idea of using the sandboxes as a means of tracking if the pod is expected to be running. Are there situations where we need to check both sandbox and containers? e.g. I wouldn't expect a container to ever be running if the sandbox was not (networking wouldn't work) - at least in the case of docker (no idea for other runtimes).

Reply:
Overall I don't expect that this would cause a problem - just curious if this is in fact a situation we might hit. cc @yifan-gu - any thoughts here?

Reply:
This behavior of checking both sandboxes and containers is how the kubelet performs the GetPods() action that was the implementation of the /runningpods endpoint. So I kept to to prevent changes to behavior

Reply:
From in-person discussion: possibly redundant - but this is safer regardless (and what the kubelet does - so may as well do the same).

Reply:
yeah, checking both here is more accurate as there can be a moment that the "sandbox" is down, but the containers are still running. They will be killed by kubelet shortly after kubelet finds the sandbox is down. Then kubelet will restart the whole pod (sandbox + containers).
And I think that's not the case we want to activate the checkpoint since kubelet will try to restart it. So I agree with keeping as is here.